src/coverlet.core/Symbols/CecilSymbolHelper.cs

```diff
@@ -396,6 +412,256 @@ private bool SkipGeneratedBranchesForExceptionHandlers(MethodDefinition methodDe
+            // by the compiler.  We want to skip the last three, but we want to
+            // keep the first one.
+            //
+            // (1) At each iteration of the loop, a check that there is another
+            //     item in the sequence.  This is a branch that we want to keep,
+            //     because it's basically "should we stay in the loop or not?",
+            //     which is germane to code coverage testing.
+            // (2) A check near the end for whether the IAsyncEnumerator was ever
+            //     obtained, so it can be disposed.
+            // (3) A check for whether an exception was thrown in the most recent
+            //     loop iteration.
+            // (4) A check for whether the exception thrown in the most recent
+            //     loop iteration has (at least) the type System.Exception.
+            //
+            // If we're looking at any of the last three of those four branches,
+            // we should be skipping it.
+
+            int currentIndex = instructions.BinarySearch(instruction, new InstructionByOffsetComparer());
+
+            return SkipGeneratedBranchForAwaitForeach_CheckForAsyncEnumerator(instructions, instruction, currentIndex) ||
+                   SkipGeneratedBranchForAwaitForeach_CheckIfExceptionThrown(instructions, instruction, currentIndex) ||
+                   SkipGeneratedBranchForAwaitForeach_CheckThrownExceptionType(instructions, instruction, currentIndex);
+        }
+
+        // The pattern for the "should we stay in the loop or not?", which we don't
+        // want to skip (so we have no method to try to find it), looks like this:
+        //
+        // IL_0111: ldloca.s 4
+        // IL_0113: call instance !0 valuetype [System.Private.CoreLib]System.Runtime.CompilerServices.ValueTaskAwaiter`1<bool>::GetResult()
+        // IL_0118: brtrue IL_0047
+        //
+        // In Debug mode, there are additional things that happen in between
+        // the "call" and branch, but it's the same idea either way: branch
+        // if GetResult() returned true.
+
+        private bool SkipGeneratedBranchForAwaitForeach_CheckForAsyncEnumerator(List<Instruction> instructions, Instruction instruction, int currentIndex)
+        {
+            // We're looking for the following pattern, which checks whether a
+            // compiler-generated field of type IAsyncEnumerator<> is null.
+            //
+            // IL_012c: ldfld class [System.Private.CoreLib]System.Collections.Generic.IAsyncEnumerator`1<int32> AwaitForeachStateMachine/'<AsyncAwait>d__0'::'<>7__wrap1'
+            // IL_0131: brfalse.s IL_0196
+
+            if (instruction.OpCode != OpCodes.Brfalse &&
+                instruction.OpCode != OpCodes.Brfalse_S)
+            {
+                return false;
+            }
+
+            if (currentIndex >= 1 &&
+                instructions[currentIndex - 1].OpCode == OpCodes.Ldfld &&
```

Comment:
Can we add the `ldarg` before `ldfld` and `brfalse.s` to be more sure?

![image](https://user-images.githubusercontent.com/7894084/110209809-ca55ab80-7e8e-11eb-996b-c5ec48b20fd5.png)

Reply:
Yep.  Will do.